### PR TITLE
Integrated Async Java Client

### PR DESCRIPTION
**Pull request by owaiskazi19**
### Description
To invoke the Actions residing in OpenSearch. Extensions will use Java Client to call the action using REST requests available. Currently, we support synchronous Java Client. As AD plugin uses asynchronous request calls, create an async Java Client for the same.

This is needed for the migration of [Start Detector](https://github.com/opensearch-project/opensearch-sdk-java/issues/383).

### Issues Resolved
Fixes https://github.com/opensearch-project/opensearch-sdk-java/issues/557
Fixes https://github.com/opensearch-project/opensearch-sdk-java/issues/296

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/sdk/SDKClient.java [hunk 1]
```diff
@@ -147,6 +172,17 @@ public OpenSearchClient initializeJavaClient(ExtensionSettings settings) {
         return initializeJavaClient(settings.getOpensearchAddress(), Integer.parseInt(settings.getOpensearchPort()));
     }
 
+    /**
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/SDKClient.java:175]: (Style) Please pair the overloaded client initialize methods together. See https://google.github.io/styleguide/javaguide.html#s3.4.2-ordering-class-contents

Specifically, move this method below `initializeJavaClient()`.

File: src/test/java/org/opensearch/sdk/TestSDKClient.java [hunk 2]
```diff
@@ -63,6 +66,15 @@ public void testCreateJavaClient() throws Exception {
         sdkClient.doCloseJavaClient();
     }
 
+    @Test
+    public void testCreateJavaAsyncClient() throws Exception {
+        OpenSearchAsyncClient javaAsyncClient = sdkClient.initializeJavaAsyncClient(settings);
+        assertInstanceOf(OpenSearchIndicesAsyncClient.class, javaAsyncClient.indices());
+        assertInstanceOf(OpenSearchClusterAsyncClient.class, javaAsyncClient.cluster());
+
+        sdkClient.doCloseJavaClient();
```
Comment (dbwiddis) [inline src/test/java/org/opensearch/sdk/TestSDKClient.java:75]: Shouldn't this be `doCloseJavaAsyncClient()` (which doesn't yet exist but should, along with adding to `close()`?)

File: src/main/java/org/opensearch/sdk/SDKClient.java [hunk 3]
```diff
@@ -147,6 +172,17 @@ public OpenSearchClient initializeJavaClient(ExtensionSettings settings) {
...
+     * Initializes an OpenAsyncSearchClient using OpenSearch JavaClient
+     *
+     * @param settings The Extension settings
+     * @return The SDKClient implementation of OpenSearchAsyncClient. The user is responsible for calling
+     *         {@link #doCloseJavaClient()} when finished with the client
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/SDKClient.java:180]: The user may be responsible for closing this client, but this points them to the wrong client to close. :-)

**Review by dbwiddis — CHANGES_REQUESTED** [review]

File: src/test/java/org/opensearch/sdk/TestSDKClient.java (hunk 2 above)
Comment (owaiskazi19) [reply src/test/java/org/opensearch/sdk/TestSDKClient.java:75]: This is how `doCloseJavaClient` looks like:

```java
public void doCloseJavaClient() throws IOException {
        if (restClient != null) {
            restClient.close();
        }
    }
```

We are actually closing the restClient and it's the same for `JavaClient` and `JavaAsyncClient` as I have created a new method to `initalizeTransport` where we actually build the `restClient`. I'm not sure how different `doCloseJavaAsyncClient` gonna look like.

File: src/main/java/org/opensearch/sdk/SDKClient.java (hunk 3 above)
Comment (dbwiddis) [reply src/main/java/org/opensearch/sdk/SDKClient.java]: If we're using the same close method for both clients that should be documented here as well.

**Review by dbwiddis — APPROVED** [review]
LGTM for now.   The whole initialize and close thing really could use an overhaul at some point.

**Review by ryanbogan — APPROVED** [review]

**Pull request merged**